### ISSUE TITLE
Db Struct

### ISSUE THREAD

**Issue by DanEvans485** [G-Bro/OpenDM-Server#14]
This is not everything but I wanted to visualise the class and model structure. Feel free to remake in [Draw.io](https://www.draw.io/)

![image](https://user-images.githubusercontent.com/29194430/73987535-a1d04a80-4938-11ea-9bc5-8aa33c5f6461.png)

**Comment by Newobc** [comment]
This made my brain hurt a lil bit.

But seems relatively correct, only problem is, we're still noobs, so we're bound to make mistakes with this.

I'd say its going to be really important (at least from a FE perspective) to keep the naming tight, and close to convention, as the naming and jargon is one of the most confusing aspects of DnD.

Looks gud tho!

**Comment by G-Bro** [comment]
I just realised Monsters aren't characters. Monsters are more comparable to races.

If you have three orcs then they still use the same monster record in the table but have their own health.

So there has to be a separate Monster and MonsterCharacter class (where monster is the template for a monster character)

**Comment by G-Bro** [comment]
Or maybe MonsterSpecies and MonsterCharacter

**Comment by G-Bro** [comment]
@DanEvans485 are you suggesting that all Character classes use the same database table?

Same with all items?

**Comment by DanEvans485** [comment]
[Draw.io version](https://drive.google.com/file/d/11SL9Ca9twB_hb1Sq7qqkaVxXbmlZ3V5E/view)

**Comment by DanEvans485** [comment]
> @DanEvans485 are you suggesting that all Character classes use the same database table?
> 
> Same with all items?

Yes are they not the same?

**Comment by DanEvans485** [comment]
Just they use default values instead of custom ones like players

**Comment by DanEvans485** [comment]
It's gotten out of hand

**Comment by G-Bro** [comment]
@DanEvans485 I've modelled everything we need (I think) for the Monster class https://www.draw.io/#G1pUeQyoH7rnbVNscC5p2PsKyAAtNVRImX

**Comment by DanEvans485** [comment]
Cant access

**Comment by G-Bro** [comment]
https://drive.google.com/file/d/1pUeQyoH7rnbVNscC5p2PsKyAAtNVRImX/view?usp=sharing

**Comment by DanEvans485** [comment]
So you don't think having armour and weapons as items would be good?

**Comment by DanEvans485** [comment]
And I still think it'd be beneficial having monsters and characters as one table

**Comment by G-Bro** [comment]
I thought we agreed to keep item types separate, like armour, weapons, potions, rings etc.

Monsters are the classes. We can still have a single character table that has either a monster_id or a class_id and race_id

**Comment by DanEvans485** [comment]
> I thought we agreed to keep item types separate, like armour, weapons, potions, rings etc.

Separate classes same table I thought

**Comment by DanEvans485** [comment]
```
class Item
class Weapon extends Item
class Armour extends Item
```

**Comment by DanEvans485** [comment]
It's just too me they seem so similar. The only difference is Monsters are just made from a preset and never changed and characters do get changed.

**Comment by G-Bro** [comment]
Monsters are not characters though. If you have three orc characters then they belong to one 'monster'. The monster is essentially the species of the character that provides the template that the character will be made from,  so there is nothing to stop monster characters being characters.

You will have one record in the model table for 'orc' but if one orc character takes damage then that is applied to the orc character, not the monster.

As for armour, I think using item modifiers complicates things. The function below is how I calculate armour class in the current method, and I think this is massively simpler than using item modifiers with string commands

````php
    public function getArmorClass(Character $character) {
        $armor_class = $this->base_armor;

        if ($this->apply_dex_modifier) {
            if ($this->dex_cap) {
                $armor_class += min(
                    $this->dex_cap,
                    $character->dexterity_modifier
                );
            } else {
                $armor_class += $character->dexterity_modifier;
            }
        }

        return $armor_class;
    }
````

**Comment by DanEvans485** [comment]
How do you compute armour class. Like normally?

**Comment by G-Bro** [comment]
Each type of armour has a base AC

* For light armour, you add your dexterity modifier on top of that base AC
* For medium armour, you add your dexterity modifier up to a maximum of 2 to the base AC
* For heavy armour you just get the base AC, but most require a certain level of strength to equip

A shield grants you a further +2 AC

https://roll20.net/compendium/dnd5e/Armor#content

**Comment by DanEvans485** [comment]
I'm gonna try something. If it fails or is too complicated then I agree with ^^ that struct

**Comment by G-Bro** [comment]
Cool. Sounds good. In that structure I think there is still a place for your modifiers, but I think the core mechanics of certain item types should be more strictly defined. Modifiers should be used for unique attributes IMO

**Comment by DanEvans485** [comment]
My main reasoning for most of my opinions to perhaps build the core of this software in away that isn't strictly locked down to D&D 5e (If we design the core well we could have the option to expand in the future), as well as db / class structure that has minimal duplicate logic.

**Comment by G-Bro** [comment]
Tabletop games are so complex and varied though that I don't think we can make a single system that works for two distinct games, if that's what you are suggesting. I really don't like the idea of storing logic in the database either, or formulas for logic. The more I think about it the less on board I am with this modifiers plan.

**Comment by DanEvans485** [comment]
Ok up to you. Did you see what I put in slack? It's not really logic in the db more an address to some data on a model.

**Comment by DanEvans485** [comment]
If this is my data (Would be in db ofc):
![image](https://user-images.githubusercontent.com/29194430/74090024-e4ac3280-4a9e-11ea-91b6-7775f96c550f.png)

This returns 14 which should be the ac if this was medium armour
![image](https://user-images.githubusercontent.com/29194430/74090020-df4ee800-4a9e-11ea-922e-8ce1c86fe59c.png)

Where the bit in the middle would be something like this on the Modifier model
https://i.imgur.com/HNZ10UU.png
![image](https://user-images.githubusercontent.com/29194430/74090016-d9f19d80-4a9e-11ea-8bad-717555dc0a1e.png)

More validation ofc

**Comment by G-Bro** [comment]
My alternative proposal would work like so:

Say you have a ring that grants +1 AC (ring of protection)

You could have a ring table, and a stats table. A join table from ring to stats would contain a pivot field called bonus.

So in this case the ring would have a structure along the lines of:

````php
[
    'name' => 'Ring of Protection',
    'stat_modifiers' => [
        0 => [
            'code' => 'armor_class',
            'pivot' => [
                'bonus' => 1
            ]
        ]
    ]
]
````

then to apply those modifiers:

````php
foreach ($character->rings as $ring) {
    foreach ($ring->stat_modifiers as $modifier) {
        $character->[$modifier->code] += $modifier->pivot->bonus;
    }
}
````

**Comment by DanEvans485** [comment]
And in my case it'd look something like:
```php
[
   'name' => 'Ring of Protection',
   'type' => 'ring',
   'modifiers' => [
     [
       'type' => 'stat',
       'name' => 'ac',
       'value' => 1,
     ]
   ],
]
```

Then to apply you'd do
```php
function getAc() {
    return $this->base_ac + $this->active_items
        ->pluck('modifiers')
        ->where('type', 'stat')
        ->where('name', 'ac')
        ->sum('value');
}
```

**Comment by G-Bro** [comment]
Yeah, looks like we had very similar plans after all. What would you propose the DB structure to be in your system? Because you have 'value' which can be numeric or a dot-notation string.

Perhaps it could be something like:

| id | skill_id | modifier | skill_modifier_id | max | min |
| -- | ------ | -------- | ---------------- | ---- | ---- |
| 1 | 6 (AC) | 4 | 4 (DEX) | 14 | 0 |

so then the logic becomes:

````php
$character->[$modifer->code] += $modifier->pivot->bonus;
$character->[$modifer->code] += $character->[$modifier->pivot->$skill_modifier->$code];

$character->[$modifier->code] = clamp(current, $modifier->min, $modifier->max);
````

because there are several instances where you apply modifiers that are +(n) + (ability modifier)

**Comment by DanEvans485** [comment]
The cases where it relies on x attribute of the character I used to dot notation to define it. And like in the armour case I posted above you have two mods, one applies the base ac and the second applies the dex bonus. To me it seems like it worked but I can't be sure.

**Comment by DanEvans485** [comment]
So in a case where that ring would apply a base ac of 1 and a bonus of the characters dex to a max of 4, it'd look like so
```php
[
   'name' => 'Ring of Protection',
   'type' => 'ring',
   'modifiers' => [
     [
       'type' => 'stat',
       'name' => 'ac',
       'value' => 1,
     ],
     [
       'type' => 'stat',
       'name' => 'ac',
       'value' => 'character.skills.dex',
       'max' => 4,
     ]
   ],
]
```

Then to apply you'd do
```php
function getAc() {
    return $this->base_ac + $this->active_items
        ->pluck('modifiers')
        ->where('type', 'stat')
        ->where('name', 'ac')
        ->sum('value');
}
```

**Comment by Newobc** [comment]
When your side project looks more legit than your salaried job...

**Comment by DanEvans485** [comment]
Yeah how much are you paying us @G-Bro ?

**Comment by G-Bro** [comment]
I might let you use it
